pup-0003.md

```diff
@@ -0,0 +1,154 @@
+   branches the change has been made to.
+
+### Hotfix
+
+The current hotfix process [2] would remain basically unchanged. However,
+instead of merging forward the commit, we'd use cherry-picking.
+
+1. The affected release would be branched and a hotfix would be applied.
+2. After the release, the branch gets merged into the x.y-release branch.
+3. The change is then cherry-picked into the x.y-dev branch and then
+   cherry-picked forward to any other x.y-dev branches and master.
+4. The person cherry-picking would also need to track in redmine which release
+   branches the change has been made to.
+
+### Updating redmine
+
+For tracking cherry-picks in redmine, a few proposals include:
+
+* Use a second/different redmine issue to track the inclusion of the backport
+into a given release and associate the two somehow. Maybe a custom field called
+'cherry-picks' could record issue numbers as a comma separated list which would
+be clickable into a redmine query.
+* Use an associated commit. With this a redmine issue will show the original
+commits and any cherry-picks. We could have a new commit keyword so a commit
+could contain "cherrypick #1234" which would associate the cherry-pick commit
+with issue 1234. It would not modify the state of the issue so whatever state
+it's in, e.g. CLOSED - CURRENT RELEASE or MODIFIED would remain as-is.
+* Use a custom field to track the cherry-picks. This is convenient for release
+engineering queries, but would not show the actual cherry-pick commits
+themselves.
+
+
+## Drawbacks
+
+Git is really built for changes to be merged between branches. For example,
+when cherry-picking a change, the commit is assigned a new id but when merging
+the commit, it retains the same id. It makes looking up the commit to see what
+branches it's been applied to easier. If we cherry-pick, we have to track this
+information externally to git. This means also having to remember to update
+redmine.
+
+Switching to this model also creates some inconsistency. For example, if we
+adopt this for Pulp 3, then we still have to use merge forwarding for Pulp 2.
+Then we are stuck having to support two different strategies.
+
+Also, when creating bug fixes or hotfixes in the cherry-pick model, code will
+be developed against master. This could cause problems when the code is cherry
+picked to release branches such as the change relying on code that is in master
+but not in the release branch. This requires extra scrutiny and awareness on
+the part of the person cherry-picking a change back, to ensure that all code
+touched or used by a change exists on the release branch and behaves the same.
+Sometimes it will require cherry-picking intermediate changes that might not
+otherwise be obvious.
+
+Defaulting to master, and requiring extra steps to get a change onto the
+current release branch, is likely to result in fewer bug fixes making it onto
+current release branches. We can and should try to be diligent about
```

Comment:
As an alternative to this downside, one person could cherry pick everything backwards regularly. For example once a week.